src/core/geometry/qgscircularstring.cpp

```diff
@@ -1592,6 +1594,118 @@ double QgsCircularString::segmentLength( QgsVertexId startVertex ) const
   return QgsGeometryUtilsBase::circleLength( x1, y1, x2, y2, x3, y3 );
 }
 
+double QgsCircularString::distanceBetweenVertices( QgsVertexId fromVertex, QgsVertexId toVertex ) const
+{
+  // Convert QgsVertexId to simple vertex numbers for curves (single ring, single part)
+  if ( fromVertex.part != 0 || fromVertex.ring != 0 || toVertex.part != 0 || toVertex.ring != 0 )
+    return -1.0;
+
+  int fromVertexNumber = fromVertex.vertex;
+  int toVertexNumber = toVertex.vertex;
+
+  if ( fromVertexNumber < 0 || fromVertexNumber >= numPoints() || toVertexNumber < 0 || toVertexNumber >= numPoints() )
```

Comment:
retrieve numPoints once upfront and store in a local int -- it'll be cheaper than the multiple calls, as those involve a std::min calculation.